analytics/input_redactor.go

```diff
@@ -0,0 +1,273 @@
+	v := value.String()
+	e, err := i.expand(options, v)
+	if err != nil {
+		return "", "", err
+	}
+	v, err = RedactStringWithSecret(v, i.secrets)
+	if err != nil {
+		return "", "", err
+	}
+	e, err = RedactStringWithSecret(e, i.secrets)
+	if err != nil {
+		return "", "", err
+	}
+	return v, e, nil
+}
+
+func (i inputRedactor) expand(options models.EnvironmentItemOptionsModel, value string) (string, error) {
+	result := value
+	var envs map[string]string
+	if options.IsTemplate != nil && *options.IsTemplate {
+		if envs == nil {
+			effects, err := env.GetDeclarationsSideEffects(i.environment, i.envSource)
+			if err != nil {
+				return "", err
+			}
+			envs = effects.ResultEnvironment
+		}
+
+		evaluatedValue, err := bitrise.EvaluateTemplateToString(result, i.isCIMode, i.isPullRequestMode, i.buildRunResults, envs)
+		if err == nil {
+			result = evaluatedValue
+		}
+	}
+	if options.IsExpand != nil && *options.IsExpand {
+		if envs == nil {
+			effects, err := env.GetDeclarationsSideEffects(i.environment, i.envSource)
+			if err != nil {
+				return "", err
+			}
+			envs = effects.ResultEnvironment
+		}
+		result = os.Expand(result, func(s string) string {
+			if _, ok := envs[s]; !ok {
+				return ""
+			}
+			return envs[s]
+		})
+	}
+	return result, nil
+}
+
+func getData(input models.EnvironmentItemModel) (models.EnvironmentItemOptionsModel, string, interface{}, error) {
+	options, err := input.GetOptions()
+	if err != nil {
+		return models.EnvironmentItemOptionsModel{}, "", nil, err
+	}
+	var keys []string
+	var values []interface{}
+
+	for key, value := range input {
+		keys = append(keys, key)
+		values = append(values, value)
+	}
+
+	if len(keys) == 0 {
+		return models.EnvironmentItemOptionsModel{}, "", nil, errors.New("no environment key specified")
+	} else if len(keys) > 2 {
+		sort.Strings(keys)
+		return models.EnvironmentItemOptionsModel{}, "", nil, fmt.Errorf("more than 2 keys specified: %v", keys)
+	}
+
+	// Collect env key and value
+	key := ""
+	var value interface{}
+	optionsFound := false
+
+	for i := 0; i < len(keys); i++ {
+		k := keys[i]
+		if k != models.OptionsKey {
+			key = k
+			value = values[i]
+		} else {
+			optionsFound = true
+		}
+	}
+
+	if key == "" {
+		sort.Strings(keys)
+		return models.EnvironmentItemOptionsModel{}, "", nil, fmt.Errorf("no environment key found, keys: %v", keys)
+	}
+	if len(keys) > 1 && !optionsFound {
+		sort.Strings(keys)
+		return models.EnvironmentItemOptionsModel{}, "", nil, fmt.Errorf("more than 1 environment key specified: %v", keys)
+	}
+	return options, key, value, nil
+}
+
+// RedactStringWithSecret ...
+func RedactStringWithSecret(value string, secrets []string) (string, error) {
+	src := bytes.NewReader([]byte(value))
+	dstBuf := new(bytes.Buffer)
+	secretFilterDst := filterwriter.New(secrets, dstBuf)
+
+	if _, err := io.Copy(secretFilterDst, src); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, stream copy failed: %s", err)
```

Comment:
Since this is secret filtering, I would be extra cautious here and wouldn't include the arbitrary error in the error message (that will get printed to the build log eventually). We can't rule out that the error message will contain the secret values.

Reply:
A.k.a. who is redacting the error of the input redactor? 😆